libs/langchain/langchain/chat_models/base.py

```diff
@@ -200,3 +342,124 @@ def _check_pkg(pkg: str) -> None:
             f"Unable to import {pkg_kebab}. Please install with "
             f"`pip install -U {pkg_kebab}`"
         )
+
+
+def _remove_prefix(s: str, prefix: str) -> str:
+    if s.startswith(prefix):
+        s = s[len(prefix) :]
+    return s
+
+
+_CHAT_MODEL_METHODS_TAKE_CONFIG = tuple(
+    name
+    for name, func in inspect.getmembers(BaseChatModel, inspect.isfunction)
+    if "config" in inspect.signature(func).parameters
+)
+
+_DECLARATIVE_METHODS = (
+    "assign",
+    "bind",
+    "map",
+    "pick",
+    "pipe",
+    "with_alisteners",
+    "with_fallbacks",
+    "with_listeners",
+    "with_retry",
+    "with_types",
+    "bind_tools",
+    "with_structured_output",
+    "configurable_fields",
+    "configurable_alternatives",
+)
+
+
+class _ConfigurableModel:
+    def __init__(
+        self,
+        *,
+        default_config: Optional[dict] = None,
+        config_prefix: str = "",
+        configure_any: bool = False,
+        queued_operations: Sequence[Tuple[str, Tuple, Dict]] = (),
+    ) -> None:
+        self._default_config = default_config or {}
+        self._config_prefix = config_prefix
+        self._configure_any = configure_any
+        self._queued_operations: List[Tuple[str, Tuple, Dict]] = list(queued_operations)
+
+    def __getattr__(self, name: str) -> Any:
+        if name in _CHAT_MODEL_METHODS_TAKE_CONFIG:
+
+            def from_config(
+                *args: Any, config: Optional[RunnableConfig] = None, **kwargs: Any
+            ) -> Any:
+                return getattr(self._model(config), name)(
+                    *args, config=config, **kwargs
+                )
+
+            return from_config
+        elif name in _DECLARATIVE_METHODS:
+
+            def queue(*args: Any, **kwargs: Any) -> _ConfigurableModel:
+                self._queued_operations.append((name, args, kwargs))
+                return self
+
+            return queue
+        elif self._default_config and hasattr(self._model(), name):
```

Comment:
`self._model()` may be expensive to instantiate (e.g., opens network connection), would be good to remove unnecessary instantiations